lib/floki/traversal.ex

```diff
@@ -0,0 +1,23 @@
+defmodule Floki.Traversal do
+  @moduledoc false
+
+  def traverse_and_update(text, _fun) when is_binary(text), do: text
+
+  def traverse_and_update([head | tail], fun) do
+    case traverse_and_update(head, fun) do
+      nil -> traverse_and_update(tail, fun)
+      mapped_head -> [mapped_head | traverse_and_update(tail, fun)]
+    end
+  end
+
+  def traverse_and_update([], _fun), do: []
+
+  def traverse_and_update({elem, attrs, children}, fun) do
+    mapped_children = traverse_and_update(children, fun)
+    fun.({elem, attrs, mapped_children})
+  end
+
+  def traverse_and_update({:comment, children}, fun) do
```

Comment:
we can have some other cases like `{:pi, _, _}` and `{:pi, _}`. So I think you could have a clause similar to the first one, but in the bottom.

```elixir
def traverse_and_update(html_node, _fun), do: html_node
```

Reply:
I'm going to fix this after merge.